### PR TITLE
return proper length if contentLength is known

### LINKED ISSUE #26735: [ManagedHandler] HttpConnectionContent.TryComputeLength() always returns false/0
The function should return true and proper value if Content-length has been specified in response. 
Aside from debugging this may help with decision if to drain certain responses or to abort connection.

### PR DESCRIPTION
**Pull request by wfurt**
fixes #26735

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionContent.cs [hunk 1]
```diff
@@ -54,6 +54,11 @@ private HttpContentStream ConsumeStream()
 
             protected internal override bool TryComputeLength(out long length)
             {
+                if (Headers.ContentLength != null)
+                {
+                    length = Headers.ContentLength.GetValueOrDefault();
+                    return true;
+                }
                 length = 0;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionContent.cs]: Nit: blank line after the brace

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionContent.cs [hunk 2]
```diff
@@ -54,6 +54,11 @@ private HttpContentStream ConsumeStream()
 
             protected internal override bool TryComputeLength(out long length)
             {
+                if (Headers.ContentLength != null)
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionContent.cs]: ContentLength isn't a trivial property, so its result should be stored into a local, e.g.
```C#
long? contentLength = Headers.ContentLength;
if (contentLength.HasValue)
{
    length = contentLength.GetValueOrDefault();
    return true;
}

length = 0;
return false;
```

**Comment by stephentoub** [comment]
@wfurt, you labeled the issue as a bug, but I'm not convinced of that.  I don't see any other handlers doing the equivalent here.

@davidsh, is this a good change?

**Comment by davidsh** [comment]
>@davidsh, is this a good change?

I don't understand this change.  What is the purpose of the HttpConnectionContentClass? Does it represent the response of the HTTP request?

The response of the HTTP request itself is a network based stream.  So, there should be an class that provides Stream class interface to that.  For WinHttpHandler, that is WinHttpResponseStream.

However, HttpMessageHandler.SendAsync() returns an HttpResponseMessage.  And within that there is a .Content property to represent the HttpContent object.  Usually, we create the response of the HTTP request as a System.Net.Http.StreamContent object and pass in the appropriate Stream-derived class from that.  As such, we do not have to do anything with TryComputeLength() because it is supposed to use the semantics of what StreamContent use.

I'm not sure if this HttpConnectionContent class is some hybrid of the above concepts.  But it is atypical way of implementing a protocol handler and response content / stream.

**Comment by geoffkizer** [comment]
> What is the purpose of the HttpConnectionContentClass? Does it represent the response of the HTTP request?

Yes.

> The response of the HTTP request itself is a network based stream. So, there should be an class that provides Stream class interface to that. For WinHttpHandler, that is WinHttpResponseStream.

There is; actually there are several: ContentLengthReadStream, ChunkedEncodingReadStream, etc.

> Usually, we create the response of the HTTP request as a System.Net.Http.StreamContent object and pass in the appropriate Stream-derived class from that. As such, we do not have to do anything with TryComputeLength() because it is supposed to use the semantics of what StreamContent use.

Agreed, this would be better to do.  The existing HttpConnectionContent is essentially just StreamContent.

The reason HttpConnectionContent exists, if I recall correctly, has to do with the order in which things are constructed when we read a response from the wire.  We parse status line, then headers, then based on the header values we construct an appropriate Stream for the response body.  The problem here is that when parsing the headers, we need to have the Content object already, so that we can set content headers into it.  But at that point we don't have the stream yet, so we can't actually construct StreamContent, since the only way to give it the Stream is in its constructor.

So the only real thing that HttpConnectionContent does that StreamContent doesn't, is it allows you to call SetStream after construction.

This seems unnecessarly complicated and duplicative, but I'm not sure what the best fix is.  Thoughts?

**Comment by geoffkizer** [comment]
Also, when making changes like this, please add tests.  This has the added benefit of confirming we are matching behavior of existing handlers.

**Comment by stephentoub** [comment]
@wfurt, I'm going to close this.  Per the discussion here and my comments in https://github.com/dotnet/corefx/issues/26735, I don't believe this is needed.

**Pull request closed without merging**